File: src/main/java/org/veupathdb/lib/gradle/container/util/FileFilter.java
package org.veupathdb.lib.gradle.container.util;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.function.Predicate;

@FunctionalInterface
public interface FileFilter extends Predicate<File> {

  @NotNull
  static FileFilter contains(@NotNull String substring) {
    return f -> f.getName().contains(substring);
  }

  @NotNull
  static FileFilter endsWith(@NotNull String suffix) {
    return f -> f.getName().endsWith(suffix);
  }

  static FileFilter beginsWith(@NotNull String prefix) {
    return f -> f.getName().startsWith(prefix);
  }

  @NotNull
  default FileFilter and(@NotNull FileFilter other) {
    return f -> test(f) && other.test(f);
  }
}

File: src/main/java/org/veupathdb/lib/gradle/container/util/FileFilter.kt
package org.veupathdb.lib.gradle.container.util

import java.io.File
import java.util.function.Predicate

fun interface FileFilter : Predicate<File> {

  companion object {
    @JvmStatic
    fun contains(substring: String): FileFilter {
      return FileFilter { it.name.contains(substring) }
    }

    @JvmStatic
    fun endsWith(suffix: String): FileFilter {
      return FileFilter { it.name.endsWith(suffix) }
    }

    @JvmStatic
    fun beginsWith(prefix: String): FileFilter {
      return FileFilter { it.name.startsWith(prefix) }
    }
  }

  fun and(other: FileFilter): FileFilter {
    return FileFilter { test(it) && other.test(it) }
  }
}
